# `tigre clean`

`tigre clean` performs several preparatory steps to standardize the GFF3 files for downstream analyses.

Two fundamental preparatory steps are handling bondaries of circular genomes, and resolving overlapping features. These steps are essential because overlapping annotations and/or circular genomes can make it tricky (computationally) to identify and extract intergenic regions.

## Features that span the genome boundary

Sometimes features span the "boundary" of a circular genome, meaning they begin before the genome end coordinate, and extend beyond the genome start coordinate.
These boundary-spanning features present a computational challenge for downstream analyses.

In such cases, `tigre clean` breaks these features into two parts: one at the end of the genome and one at the start genome, appending their type with "_fragment". This ensures that intergenic regions are correctly identified even when features cross the genome boundary.

To resolve this challenge, tigre clean splits each boundary-spanning feature into two separate annotations, appending "_fragment" to their GFF3 feature type. The splitting is done as follows:

 - First fragment: Spans from the original feature start coordinate until the end of the genome
 - Second fragment: Spands from the genome start (coordinate 1) to the original feature end position minus the genome size

Both fragments retain all the original feature's attributes. This splitting strategy simpliflies subsequent steps by allowing them to initially treat a circular genome as linear, and dealing with the circularity only when processing regions around the genome boundary.

## Overlapping Features

If `tigre clean` detects that multiple features overlap one another, the `clean` command creates a new feature type called `overlapping_feature_set`. This special feature marks regions where no valid intergenic sequence can exist due to feature overlap.

The `overlapping_feature_set` is generated by merging all overlapping features into one consolidated annotation. Its strand is always set to `+`, since it represents a combined (meta)region rather than a feature on a specific strand.

To preserve context for downstream analyses, the `overlapping_feature_set` stores the identifying information (i.e., gene name) only of the features at the boundaries of the overlapping region. These boundary features are used to determine the upstream and downstream delimiters of the adjacent intergenic regions.

The attributes of the `overlapping_feature_set` include:
 - `name_left`: Retrieved from the `ID=` value in the attribute column (or the standardized GDT gene name) of the feature that starts at the left boundary of the overlap region.
 - `source_left`: Custom value built using the original feature information in the following format: `seqid|type|start|end|strand|ID=;`
 - `name_right`: Retrieved from the `ID=` value in the attribute column (or the standardized GDT gene name) of the feature that ends at the right boundary of the overlap region.
 - `source_right`: Custom value built using the original feature information in the following format: `seqid|type|start|end|strand|ID=;`

If there are multiple features that start or end at the overlap region boundaries, `tigre clean` uses the following rules to select which features to pick as left and right boundary features:

- Left: Choose the feature with the lowest start coordinate. If there is a tie, select the feature with the highest end coordinate.
- Right: Choose the feature with the highest end coordinate. If there is a tie, select the feature with the lowest start coordinate.

If two features have identical start and end coordinates, the one that appears first in the GFF3 file is chosen.

There are a few possible scenarios for overlapping features, described in the image below:

<div align="center">
  <img src="img/overlap.png" />
</div>

